library/network_connections.py

```diff
@@ -1680,12 +1712,12 @@ def run(self):
                             self.run_action_down(idx)
                         else:
                             assert False
-                except Exception as e:
-                    self.log_warn(
-                        idx, "failure: %s [[%s]]" % (e, traceback.format_exc())
-                    )
+                except Exception as error:
+                    self.rollback_transaction(idx, action, error)
                     raise
 
+        self.finish_transaction()
```

Comment:
I would move the finish call inside the loop and only call it for `CheckMode.REAL_RUN`, so that it mirrors the call to start.

It makes no difference in practice, because `start_transaction()` is called zero or one times, and calling `finish_transaction()` is correct in both cases. But still, I think from the code it should be obvious that every start is exactly followed by one rollback/finish.

Reply:
changed this